### DISCUSSION TITLE
WebAuthn Authentication page UI design

### DISCUSSION THREAD

**Discussion by mabartos** [keycloak/keycloak#9038] [Ideas]
Hello!

I'm currently working on WebAuthn tests and some minor improvements and I've noticed our authentication login page for WebAuthn and I don't really like it much. Regardless the count of registered possible authenticators, there's only one button for executing authentication process, as you can see in the figure below. When the user clicks the button, credentials all those authenticators are provided to the WebAuthn API as `allowCredentials` and all those credentials can be used within authentication process. What about to leave the decision up to the user, who can choose the authenticator by his/her choice? It'd be also nice to get some information about the authenticator, which will be used.

Current page:
<img src="https://user-images.githubusercontent.com/38039883/145175928-97265555-3357-4433-bd14-4ed6c781d672.png" width="400" />

Proposed changes with one authenticator with button (it's not possible to click on the authenticator; only for information about it):
<img src="https://user-images.githubusercontent.com/38039883/145176155-1a687601-3d24-43d3-8b1a-6dc7b2258b83.png" width="400" />

Proposed changes with one authenticator without button:
<img src="https://user-images.githubusercontent.com/38039883/145176311-69fc4472-a706-40fb-8efc-0569deb8c1e9.png" width="400" />

Proposed changes for more authenticators:
<img src="https://user-images.githubusercontent.com/38039883/145176402-9d6e4601-93fe-4f45-b23e-677d596de78a.png" width="400" />

Those design are only for consideration and someone from UI team would take a look at it.
@ssilvert @mposolda WDYT?

**Comment by mposolda** [comment]
I have some impression that when we added initial WebAuthn support to the Keycloak, it worked in a similar way, which you propose in this design. Despite the fact that there were checkboxes and user was able to select one ore more WebAuthn keys. AFAIR we decided to skip that and check all the WebAuthn keys due the better usability for the user...

IMO it can be useful if user can see all the available WebAuthn authenticators. I am not 100% sure if it is good to limit him to be able to select just one of them. However I am not usability expert and not sure what is best and most intuitive for the users...

@tnorimat @stianst WDYT?

**Reply by stianst** [reply]
If we can check all webauthn authenticators and there's no negative side-effects than it seems it should stay that way?

**Reply by stianst** [reply]
Actually just as I typed that I realised that the user has to take an action for the check to happen in the first place, so I'm kinda on the fence, perhaps leaning towards what @mabartos is suggesting. I definitively hate the checkboxes idea though ;)

**Comment by Pepo48** [comment]
@mabartos I would also choose the same design that we created for the auth selector screen in order to make UX as uniform as possible. I remember that we got inspired by the official PatternFly4 design proposals, which seems to remain the same since then: 
![image](https://user-images.githubusercontent.com/12138171/145990356-3a565e6a-adad-4bb5-a46e-7663ef17a333.png)

Therefore +1 for the proposed designs. Thanks Martin!

**Comment by tnorimat** [comment]
@mabartos Basically, I agree with your UI proposal. As @mposolda mentioned, I'm also not sure which is better, allowing a user to choose only one authenticator or multiple authenticators.
My proposal is as follows.
- Allow a user to choose only one authenticator.
- Allow a user to choose all authenticator (the same as the current keycloak does)

For users registered multiple authenticators, I'm afraid that there is the case that the user cannot tell which authenticator displayed on keycloak's UI is the one the user tries to use.
To not bother such the user, it might be good to allow the user to select all authenticators.

**Reply by mposolda** [reply]
IMO the user label may help to differentiate between multiple WebAuthn keys? It should be the user responsibility to use the nice label, so he can differentiate between multiple WebAuthn keys (EG. "My Yubikey" VS "My mobile key")

However I agree it can be good that user will be able to select either one authenticator or all of them. I am not sure how to do that, so the screen won't become over-complicated to understand... We can have the button at the bottom of the screen like "Sign in with any Security Key", but not sure if that is good regarding usability and won't cause the screen to become "too complex" for the average user...

**Reply by tnorimat** [reply]
Yes, we need to avoid "too complex" UI. It might be preferable to take the not complex UI from average user.

**Comment by mabartos** [comment]
@stianst @mposolda @tnorimat @Pepo48 Thanks guys for your valuable points. 

Just one another approach for that.
What about to display, to the user, label 'Specify the Key' (in a same way like 'Try another way'), when there are registered more authenticators? The user can simply specify the required authenticator by his/her choice then. Or change the button text to 'Sign in with any Security Key'? But as was mentioned, it might cause more complexity of the screen.

<img src="https://user-images.githubusercontent.com/38039883/146525240-895e152c-dcc7-41a6-b139-d30c9230a035.png" width="400" />
When user clicks the 'Specify the Key':
<img src="https://user-images.githubusercontent.com/38039883/146525332-979d531c-f089-46bd-a3d2-59f6ea9c7ec4.png" width="400" />

And when there's only one registered authenticator, show the info about the authenticator:
<img src="https://user-images.githubusercontent.com/38039883/146525460-50c434cd-ca23-4a51-997b-96b28ff119ad.png" width="400" />

**Reply by mposolda** [reply]
In case user has also OTP, he will see 2 links `Specify the Key` and `Try another way` . Is not 2 links too much "confusion" for the average user? :-)

I personally more like your original proposal. Or the alternative when WebAuthn keys are shown by default with their labels, but just for the informative purpose with some header like "Available Security Keys" . User won't be able to click on the individual authenticators, but only on the button `Sign in with Security Key`, which will try all of them.

**Reply by mabartos** [reply]
@mposolda You're right. It's not the right way, I forgot about the alternate executions. 

Regarding the 'Available Security Keys', the functionality is now included in the WebAuthn Error page in this form:
<img src="https://user-images.githubusercontent.com/38039883/146532981-cfdc9153-53cc-4087-9a2f-bbf0862e5cb6.png" width="400" />

Is it the thing, what you meant?
Yep, it'd be great to inform user about the available authenticators. So, in summary, nothing major will change, just show those authenticators to the user. And maybe discuss the design for that a little bit.

**Reply by mposolda** [reply]
Yes, I meant something like this, but will be nice is design includes the nice "Key" pictures and "Created" label as shown in your screenshots.

**Reply by mabartos** [reply]
@mposolda Agree. Just nitpick for consideration: It'd be good to solve an issue with multiple authenticators (like two cols or sth; maybe like for IdPs).

**Reply by mposolda** [reply]
@mabartos Yes, however not sure about priority of this. I guess typical user would have 1-2 authenticators or 3 at max? More than 3 is probably very rare. But maybe I am wrong here?

**Reply by mabartos** [reply]
@mposolda Yep, it's just a corner case for that. But, it'd be good to bear it in mind for the future. 

So, I think what you've suggested could be a good approach. We can just wait for the others what they think.  

@mposolda Thanks!

**Comment by mabartos** [comment]
PR for this: #9252
